core/lib/node_framework_derive/src/macro_impl.rs

```diff
@@ -0,0 +1,190 @@
+    FromContext,
+    IntoContext,
+}
+
+impl fmt::Display for MacroKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::FromContext => write!(f, "FromContext"),
+            Self::IntoContext => write!(f, "IntoContext"),
+        }
+    }
+}
+
+/// Parser and renderer for `FromContext` derive macro.
+#[derive(Debug)]
+pub struct MacroImpl {
+    macro_kind: MacroKind,
+    ctx: CtxLabel,
+    ident: syn::Ident,
+    fields: Vec<Field>,
+}
+
+impl MacroImpl {
+    pub(crate) fn parse(macro_kind: MacroKind, input: DeriveInput) -> Result<Self> {
+        let ctx = CtxLabel::parse(&input.attrs)?.unwrap_or_default();
+        let ident = input.ident;
+        if !input.generics.params.is_empty() {
+            return Err(syn::Error::new(
+                ident.span(),
+                format!("Generics are not supported for `{macro_kind}`"),
+            ));
+        }
+        let fields = match input.data {
+            syn::Data::Struct(data) => match data.fields {
+                syn::Fields::Named(fields) => fields
+                    .named
+                    .into_iter()
+                    .map(|field| {
+                        let ident = field.ident.unwrap();
+                        let ty = field.ty;
+                        let label = CtxLabel::parse(&field.attrs)?.unwrap_or_default();
+                        Ok(Field { ident, ty, label })
+                    })
+                    .collect::<Result<Vec<_>>>()?,
+                _ => {
+                    return Err(syn::Error::new(
+                        ident.span(),
+                        format!("Only named fields are supported for `{macro_kind}`"),
+                    ))
+                }
+            },
+            _ => {
+                return Err(syn::Error::new(
+                    ident.span(),
+                    format!("Only structures are supported for `{macro_kind}`"),
+                ))
+            }
+        };
+
+        Ok(Self {
+            macro_kind,
+            ctx,
+            ident,
+            fields,
+        })
+    }
+
+    pub fn render(self) -> Result<proc_macro2::TokenStream> {
+        match self.macro_kind {
+            MacroKind::FromContext => self.render_from_context(),
+            MacroKind::IntoContext => self.render_into_context(),
+        }
+    }
+
+    fn crate_path(&self) -> proc_macro2::TokenStream {
+        if let Some(krate) = &self.ctx.krate {
+            quote! { #krate }
+        } else {
+            quote! { zksync_node_framework }
+        }
+    }
+
+    fn render_from_context(self) -> Result<proc_macro2::TokenStream> {
+        let crate_path = self.crate_path();
+        let ident = self.ident;
+        let mut fields = Vec::new();
+        for field in self.fields {
+            let ty = field.ty;
+            let ident = field.ident;
+            let default = field.label.default;
+
+            if field.label.krate.is_some() {
```

Comment:
Bikeshedding: IMO, it would be beneficial to have separate struct or at least parsers for differing contexts (despite them all using `#[context(_)]` wrapper. Using a single parser and post-checking fields like this looks not idiomatic.